include/realtime_tools/realtime_barrier.hpp

```diff
@@ -0,0 +1,468 @@
+    move.new_data_available_ = false;
+  }
+
+  // allow moving but not copying
+  MemoryBarrier(const MemoryBarrier<T> &) = delete;
+
+  MemoryBarrier & operator=(const MemoryBarrier<T> &) = delete;
+
+
+  // todo: perhaps we need a new_data on both nrt and rt side
+  // true if new data is available, depends on if the memory is being used for Read
+  // from RT or Write to RT mode
+  inline bool new_data_available() const
+  {return new_data_available_;}
+
+  void initialize(const T & value)
+  {
+    wait_lock guard(mutex_);
+    if (guard.owns_lock()) {
+      *nrt_ = *rt_ = value;
+    } else {
+      throw std::runtime_error("request to initialize realtime barrier failed trying to lock");
+    }
+  }
+
+protected:
+  T * nrt_;
+  T * rt_;
+  bool polarity_;  // flipped each time barrier rotates (swaps)
+  bool new_data_available_;
+
+  // Set as mutable so that readFromNonRT() can be performed on a const buffer
+  mutable std::mutex mutex_;
+
+  inline void swap()
+  {
+    // swap pointers
+    T * tmp = nrt_;
+    nrt_ = rt_;
+    rt_ = tmp;
+    polarity_ = !polarity_;
+  }
+
+  friend class DirectAccess<realtime>;
+
+  friend class DirectAccess<non_realtime>;
+};
+
+/// @brief Create a barrier for reading data from a realtime thread
+/// ReadBarrier implements reading data from a realtime thread using a MemoryBarrier. For example,
+/// this is used to transfer state data out of hardware interfaces. The default constructor will
+// create a new memory barrier for use. There is an alternate constructor if you want to create
+// your own MemoryBarrier explicitly.
+template<class T, class locking_strategy = wait_lock>
+class ReadBarrier
+{
+public:
+  using MemoryBarrierType = MemoryBarrier<T>;
+
+  ReadBarrier() noexcept
+  : mem_(new MemoryBarrierType()), owns_mem(true)
+  {
+  }
+
+  explicit ReadBarrier(MemoryBarrierType & mem_barrier) noexcept
+  : mem_(&mem_barrier), owns_mem(false)
+  {
+  }
+
+  virtual ~ReadBarrier()
+  {
+    if (owns_mem) {
+      delete mem_;
+    }
+  }
+
+  /// @brief Access the underlying MemoryBarrier object
+  inline MemoryBarrierType & memory()
+  {return *mem_;}
+
+  /// @brief Returns true if new data is available to read
+  inline bool new_data_available() const
+  {return mem_->new_data_available();}
+
+  /// @brief Get current value from non-realtime side without affecting the barrier.
+  /// Copy the current data into dest. No swap is performed and the new_data_available flag is
+  /// unaffected.
+  bool current(T & dest)
+  {
+    typename MemoryBarrierType::template DirectAccess<non_realtime, locking_strategy> direct(*mem_);
+    if (direct) {
+      dest = *direct;
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  /// @brief Read data out of the realtime thread
+  /// Swap RT buffer for non-RT. Copy the new data into val and reset the new_data_available flag.
+  // todo: since this read also swaps, should it be called read_and_swap() or pop() or pull()
```

Comment:
I don't think we need a rename. The consumer should always try to swap, I think that's well enough understood.